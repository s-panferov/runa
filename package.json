{
  "name": "@runy-build/monorepo",
  "type": "module",
  "license": "MIT",
  "private": true,
  "scripts": {
    "test": "NODE_OPTIONS=\"$NODE_OPTIONS --experimental-vm-modules --no-warnings\" jest",
    "check": "tsc -b"
  },
  "dependencies": {
    "@jest/globals": "^29.7.0",
    "@microsoft/api-extractor": "^7.47.12",
    "@rollup/plugin-commonjs": "^28.0.1",
    "@rollup/plugin-json": "^6.1.0",
    "@rollup/plugin-node-resolve": "^15.3.0",
    "@rsbuild/core": "^1.1.4",
    "@rspack/cli": "^1.1.3",
    "@rspack/core": "1.1.3",
    "@runy-build/runtime": "workspace:*",
    "@runy-build/schema": "workspace:*",
    "@runy-build/self": "workspace:*",
    "@runy-build/types": "workspace:*",
    "@swc/helpers": "^0.5.15",
    "@swc/jest": "^0.2.37",
    "@types/node": "^22.9.4",
    "jest": "^29.7.0",
    "rollup": "^4.27.3",
    "typescript": "5.7.2"
  },
  "devDependencies": {
    "@semantic-release/commit-analyzer": "^13.0.0",
    "@semantic-release/exec": "^6.0.3",
    "@semantic-release/github": "^11.0.1",
    "@semantic-release/release-notes-generator": "^14.0.1",
    "semantic-release": "^24.2.0",
    "semantic-release-pnpm": "^1.0.2"
  }
}
