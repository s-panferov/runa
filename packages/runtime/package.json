{
  "name": "@runy-build/runtime",
  "version": "0.1.0",
  "type": "module",
  "license": "MIT",
  "main": "./dist/index.js",
  "types": "./dist/index.d.ts",
  "dependencies": {
    "@runy-build/schema": "workspace:*",
    "execa": "^9.5.1",
    "vscode-jsonrpc": "^8.2.1",
    "winston": "^3.17.0"
  },
  "exports": {
    ".": "./dist/index.js",
    "./runy.build": "./runy.build.ts"
  },
  "files": [
    "dist/index.js",
    "dist/index.js.map",
    "dist/index.d.ts",
    "package.json",
    "README.md",
    "LICENSE"
  ]
}
