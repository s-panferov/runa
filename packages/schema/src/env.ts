import { dirname } from "node:path";
import process from "node:process";
import assert from "node:assert";

export let WORKSPACE_ROOT = "";

try {
  WORKSPACE_ROOT = process.env["BUILD_WORKSPACE"] || process.cwd();
} catch {
  // no access
}

export function setWorkspaceRoot(path: string): void {
  assert(!!path);

  // Check if the path is a URL
  try {
    const url = new URL(path);
    // If it's a valid URL, convert it to a pathname
    WORKSPACE_ROOT = url.pathname;
  } catch {
    // If it's not a valid URL, use the path as-is
    WORKSPACE_ROOT = path;
  }

  if (WORKSPACE_ROOT.endsWith(".ts")) {
    WORKSPACE_ROOT = dirname(WORKSPACE_ROOT);
  }
}
