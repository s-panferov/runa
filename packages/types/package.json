{
  "name": "@runy-build/types",
  "version": "0.1.0",
  "license": "MIT",
  "type": "module",
  "dependencies": {},
  "devDependencies": {},
  "main": "./dist/index.js",
  "types": "./dist/index.d.ts",
  "files": [
    "dist/index.js",
    "dist/index.d.ts",
    "package.json",
    "README.md",
    "LICENSE"
  ]
}
