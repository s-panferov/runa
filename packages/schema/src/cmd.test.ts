import { getPackage } from "./package";
import { BuildContext, fileset, Schema } from "./index";
import { describe, expect, test } from "@jest/globals";
import { setWorkspaceRoot } from "./env";

import { z } from "zod";

const shape = z.object({
  a: z.object({
    b: z.object({
      c: z.string(),
    }),
  }),
});

describe("output access", () => {
  test("output access", () => {
    setWorkspaceRoot(import.meta.url);

    const mod = getPackage(import.meta);

    const a = mod.target("a", (ctx) => {
      return ctx
        .in({
          files: fileset(["src/**/*"]),
        })
        .out({
          data: ctx.data(shape),
        })
        .command(
          ({ $, inp: { files }, out: { data } }) =>
            $`my command --in=${files.root} --out=${data.a.b.c}`.sh
        );
    });

    const schema = Schema.convert(a.out(new BuildContext()));
    console.log(schema);

    expect(schema).toMatchSnapshot();
  });
});
