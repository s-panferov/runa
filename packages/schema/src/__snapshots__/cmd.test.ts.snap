// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`output access output access 1`] = `
{
  "builds": {
    "dF+2veXbvQSwb9Gq8YU2Y43zYyw=": {
      "ctx": {
        "$ref": "U1wwAaqukwfarhkuFGe2QSaTdXY=",
        "kind": "context",
      },
      "spec": {
        "command": {
          "args": [
            "my command --in=",
            [
              {
                "kind": "output",
                "name": "inp",
              },
              {
                "kind": "output",
                "name": "files",
              },
              {
                "kind": "output",
                "name": "root",
              },
            ],
            " --out=",
            [
              {
                "kind": "output",
                "name": "out",
              },
              {
                "kind": "output",
                "name": "data",
              },
              {
                "kind": "output",
                "name": "a",
              },
              {
                "kind": "output",
                "name": "b",
              },
              {
                "kind": "output",
                "name": "c",
              },
            ],
          ],
          "shell": true,
        },
        "inout": {},
        "inp": {
          "files": [
            {
              "kind": "fileset",
              "path": [
                "src/**/*",
              ],
            },
          ],
        },
        "key": [],
        "out": {
          "data": [
            {
              "kind": "data",
              "schema": {
                "$schema": "http://json-schema.org/draft-07/schema#",
                "additionalProperties": false,
                "properties": {
                  "a": {
                    "additionalProperties": false,
                    "properties": {
                      "b": {
                        "additionalProperties": false,
                        "properties": {
                          "c": {
                            "type": "string",
                          },
                        },
                        "required": [
                          "c",
                        ],
                        "type": "object",
                      },
                    },
                    "required": [
                      "b",
                    ],
                    "type": "object",
                  },
                },
                "required": [
                  "a",
                ],
                "type": "object",
              },
            },
          ],
        },
        "ready": {},
      },
      "target": {
        "$ref": "XkvjJAqMr/OXcKRWsPK/hL6nSTA=",
        "kind": "target",
      },
    },
  },
  "contexts": {
    "U1wwAaqukwfarhkuFGe2QSaTdXY=": {
      "flags": {},
      "parent": undefined,
    },
  },
  "flags": {},
  "root": {
    "$ref": "dF+2veXbvQSwb9Gq8YU2Y43zYyw=",
    "kind": "build",
    "phase": "out",
  },
  "targets": {
    "XkvjJAqMr/OXcKRWsPK/hL6nSTA=": {
      "default": false,
      "name": "a",
      "package": "",
    },
  },
}
`;
